lib/src/client_credentials_grant.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:http/http.dart' as http;
+import 'package:http_parser/http_parser.dart';
+
+import 'client.dart';
+import 'handle_access_token_response.dart';
+import 'utils.dart';
+
+/// Obtains credentials using a [client credentials grant][].
+///
+/// This mode of authorization uses the client's [identifier] and [secret]
+/// to obtain an authorization token from the authorization server, instead
+/// of sending a user through a dedicated flow.
+///
+/// The client [identifier] and [secret] are required, and are
+/// used to identify and authenticate your specific OAuth2 client. These are
+/// usually global to the program using this library.
+///
```

Comment:
I think you could argue that the client credentials grant is necessary for a full implementation of OAuth2, especially since the resource owner password credentials Grant is already supported.

That being said, it is possible to put it in a separate package.

Reply:
Ah, sorry, it's actually a grant that doesn't involve a user... I thought it was a user password...

Nvm, we should probably have it for completeness :)

Reply:
Some OpenID and OAuth2 implementations need additional parameters to the request as well. Like the "aud" or "audience". So it would be important to allow the user to set these parameters into the body of the token request.

Reply:
@shaxxx, I think you might be right, there is little we need to do to land this..
But having an example would be valuable.

But you're right most of the changes left are nits, maybe we should just ship it.

Reply:
Maybe I'm wrong but seems to me changes you suggested are cosmetic. It would probably be best to pull this PR and make those changes yourself afterwards if necessary. Since this feature is critical for us we forked this branch with this PR and everything is working as expected. Obviously this is not optimal solution.